### ISSUE TITLE
Optimizer Excessive Memory Usage for Function Pointers

### ISSUE THREAD

**Issue by maximtsai** [emscripten-core/emscripten#3718]
I am currently trying to build a library containing several thousand function pointers but I am running into memory problems while optimizing JavaScript. With a simplified file of 2000 empty function pointers using -O3 optimization, optimizer.exe goes up to 12 GB of memory usage and also does not finish optimizing even after 30 minutes. I am using the latest incoming emcc and clang version 1.34.6, although this issue also happens with older versions from at least 1.29.0.

I am using a Windows 8 64-bit computer.

Link to CMake file: https://www.dropbox.com/s/5vp3pftvgjdowsc/CMakeLists.txt?dl=0
Link to simplified worker file: https://www.dropbox.com/s/oz559cljwby8yw5/EmscriptenWorker.cpp?dl=0
Link to generated .bc file: https://www.dropbox.com/s/rfehzkj7cg8u4fv/libemscripten_worker.bc?dl=0

**Comment by kripken** [comment]
Compiling that natively takes me 40 seconds and 400 MB in clang, 50 seconds and 1400 MB in gcc, so even there this simple pattern - 1,000 c++11 lambdas in a function - is surprisingly inefficient.

It's clearly even more inefficient when clang uses our libc++. We might benefit from upgrading libc++, testing that might be useful.

In general, though, the huge inefficiency even when building natively suggests that it might be best to work around this. Emitting one function per lambda, for example.

cc @sunfishcode - any idea why 1,000 c++11 lambdas would stress modern compilers so much?

**Comment by maximtsai** [comment]
Sorry, "compiling" was a misleading way to describe the issue. My compiling and linking steps finished relatively quickly with time and memory use similar to your findings, but the Javascript optimization step (optimizer.exe) was what took an inordinate amount of time and memory.

This is the last output using EMCC_DEBUG=1 (before the optimizer memory spike):

DEBUG    root:   emscript: bootstrapping struct info complete
DEBUG    root: emscript: llvm backend: D:/Emscripten/clang/fastcomp/build_incomi
ng_vs2013_64/RelWithDebInfo/bin\llc c:\users\maxim_~1\appdata\local\temp\tmpeg3h
wj\EmscriptenWorker.bc -march=js -filetype=asm -o c:\users\maxim_~1\appdata\loca
l\temp\emscripten_temp\tmpkhubwu.4.js -O3
DEBUG    root:   emscript: llvm backend took 26.0859999657 seconds
DEBUG    root: emscript: js compiler glue
DEBUG    root:   emscript: glue took 0.327999830246 seconds
DEBUG    root: asm text sizes[[234707016, 17233], 2169, 159, 3793, 0, 1624, 2378
683, 1220, 268, 2858, 2189]
DEBUG    root:   emscript: final python processing took 0.844000101089 seconds
DEBUG    root: emcc step "emscript (llvm=>js)" took 29.27 seconds
DEBUG    root: wrote memory initialization to EmscriptenWorker.js.mem
DEBUG    root: emcc step "source transforms" took 1.45 seconds
DEBUG    root: running js post-opts
DEBUG    root: applying js optimization passes: asm eliminate simplifyExpression
s simplifyIfs registerizeHarder minifyNames cleanup asmLastOpts last minifyWhite
space
chunkification: num funcs: 21986 actual num chunks: 13 chunk size range: 1811706
00 - 14932
DEBUG    root: js optimizer using native
splitting up js optimization into 13 chunks, using 8 cores  (total: 223.85 MB)

**Comment by kripken** [comment]
Ah, I didn't even let it run long enough to reach that stage, since it took minutes before I gave up. I wonder why it's so much faster for you.

In any case, I suspect that there is a single massive function emitted there, and registerizeHarder is worse than linear, so it gets stuck on that. To confirm you can look at the temp files that `EMCC_DEBUG` emits, the `emcc-1-original.js` should be before js opts. You can also run with `EMCC_DEBUG=1` to run each js opt pass separately, that might confirm registerizeHarder is the culprit.

If it is, there isn't much to do besides use -O2 instead, which should be a lot faster. registerizeHarder has been optimized quite a bit already, I'm not sure we can do better - but we know it can be slow on pathologically large functions.

**Comment by maximtsai** [comment]
Thanks for the information. I was able to get past the optimizer memory issue by breaking the single massive function containing all the function pointers into multiple smaller functions. It even solved another issue I encountered where Firefox was encountering memory spikes on startup.

**Comment by aidanhs** [comment]
This sounds interesting! I'll like to take a look, I'd hope we can do better than "it doesn't terminate".

**Comment by aidanhs** [comment]
I enabled primitive profiling on one of the (22MB) chunks from the optimizer pass of the provided bitcode:
```
# du -h preopt.js
22M     preopt.js
# time /emscripten/emsdk/emscripten/incoming_64bit_optimizer/optimizer /emscripten/slow/preopt.js asm eliminate simplifyExpressions simplifyIfs registerizeHarder minifyLocals asmLastOpts last minifyWhitespace 2>&1 >/dev/null
starting reading and parsing
    reading and parsing took 5405 milliseconds
starting asm
    asm took 0 milliseconds
starting eliminate
    EL stages: a:775892 fe:1110387 vc:136707 se:1092099 (ss:20850805) cv:731931 r:859282
    eliminate took 26610 milliseconds
starting simplifyExpressions
    simplifyExpressions took 3084 milliseconds
starting simplifyIfs
    simplifyIfs took 385 milliseconds
starting registerizeHarder
    RH stages: a:999152 fl:929272 lf:4408 bf:862980 jvua:205041 jvs:32317 jra:88812 bp:571875 r:791129
    registerizeHarder took 4590 milliseconds
starting minifyLocals
    minifyLocals took 2153 milliseconds
starting asmLastOpts
    asmLastOpts took 2050 milliseconds
starting last
    last took 0 milliseconds
starting minifyWhitespace
    minifyWhitespace took 0 milliseconds

real    0m47.313s
user    0m45.546s
sys     0m1.756s
```

Variable elimination is quite slow. Nice to see that the registerizeHarder optimisations are working -  the initial parsing before any pass is a bigger timesink!

I've got a couple of ideas I'm having a look at to speed it up.

**Comment by kripken** [comment]
Great! Thanks for looking into this.

**Comment by aidanhs** [comment]
My PR above has been merged.
Doing `EMCC_DEBUG=1 EMCC_CORES=1 emcc -O3 libemscripten_worker.bc -o /tmp/out.js` three times for before and after and recording the time from the line saying `emcc step "js opts" took X seconds`:

Before: 41.03, 40.81, 40.23
After: 28.38, 28.37, 28.40

So it's a bit bit faster now, though I'd be interested to hear how well it compiles on Windows - I wonder if the initial huge memory consumption might have been because of some poor compiler optimisations.
As an aside, using `EMCC_CORES=$NUM_LESS_THAN_NUM_CPUS` is a good way to reduce memory usage by preventing too many optimizer instances running at the same time.
